activity_browser/ui/web/tree_navigator.py

```diff
@@ -0,0 +1,602 @@
+            lca = bc.LCA(demand=fu, data_objs=data_objs)
+            lca.lci()
+            lca.lcia()
+            data = InteractiveGraphTraversal.calculate(
+                lca_object=lca, cutoff=cut_off, max_calc=max_calc, max_depth=1
+            )
+
+            # store the metadata from this calculation
+            data.metadata = {
+                "lca": lca,
+                "unit": bd.methods[method]["unit"],
+                "cut_off": cut_off,
+                "max_calc": max_calc,
+            }
+        except (ValueError, ZeroDivisionError) as e:
+            QtWidgets.QMessageBox.information(
+                None, "Nonsensical numeric result.", str(e)
+            )
+        log.debug(f"Completed graph traversal ({round(time.time() - start, 2)} seconds")
+
+        # cache the generated Graph data
+        self.cache[cache_key] = data
+
+        # generate the new Graph
+        self.graph.new_graph(data)
+        self.has_sankey = bool(self.graph.json_data)
+        self.send_json()
+
+    def set_database(self, name):
+        """Saves the currently selected database for graphing a random activity"""
+        self.selected_db = name
+
+    def random_graph(self) -> None:
+        """Show graph for a random activity in the currently loaded database."""
+        if self.selected_db:
+            method = bd.methods.random()
+            act = bd.Database(self.selected_db).random()
+            demand = {act: 1.0}
+            self.update_sankey(demand, method)
+        else:
+            QtWidgets.QMessageBox.information(
+                None, "Not possible.", "Please load a database first."
+            )
+
+    @Slot(object, name="update_graph")
+    def update_graph(self, click_dict: dict) -> None:
+        """Update the graph with the specified JSON data."""
+        traversed = self.graph.state_graph.traverse_from_node(click_dict["id"])
+        if not traversed:
+            # nothing has changed
+            return
+        self.graph.json_data = Graph.get_json_data(self.graph.state_graph.data)
+        self.send_json()
+
+
+class InteractiveGraphTraversal(NewNodeEachVisitGraphTraversal):
+    """
+    Graph traversal that maintains state for a partial depth graph traversal.
+    """
+
+    def __init__(self, nodes, edges, flows, calculation_count, metadata=None, parameters=None):
+        super().__init__()
+        self.nodes: Dict[int, GraphNode] = nodes
+        self.edges: List[GraphEdge] = edges
+        self.flows = flows
+        self.calculation_count = calculation_count
+        self.metadata = metadata or {}
+        self.parameters = parameters or {}
+        self.already_calculated = set()
+
+    @property
+    def data(self):
+        return {
+            "nodes": self.nodes,
+            "edges": self.edges,
+            "flows": self.flows,
+            "calculation_count": self.calculation_count.value,
+            "metadata": self.metadata
+        }
+
+    @classmethod
+    def calculate(
```

Comment:
I'd be a lot happier if we could incorporate missing functionality into `bw_graph_tools`; the AB has done a lot of stuff on it's own in the past, and we now have to fix the messes they created.

Reply:
moved to bw_graph_tools and cleaned it up a bit as well
https://github.com/brightway-lca/bw_graph_tools/pull/20